### PR TITLE
Fixed #182 -- split different backends escaping

### LINKED ISSUE #182: PostgreSQL search for `-` and `+`
Greetings @etianen 

Found that PostgreSQL that correctly works with `+` and `-` symbols in searching is not working in django-watson:
`RE_NON_WORD = re.compile(r'[:"(|)!><~*+-]', re.UNICODE)` which is used in `escape_query` escapes it before PG performs search.

Maybe we should have base `escape_query` and for every backend have inheritor of this method with correct backend symbols?

What do you think about this approach? If it is valid for you, I can make a PR with tests.

### PR DESCRIPTION
**Pull request by amureki**
@etianen 
So, according to #182 we wanted to separate symbols escaping for different backends.
Here is my attempt to do that.

I'm open to other suggestions/proposals.

### CODE REVIEW

**Comment by etianen** [comment]
This looks cool. I've got a few suggestions, commented inline

File: src/watson/backends.py [hunk 1]
```diff
@@ -23,25 +23,20 @@ def regex_from_word(word):
...
-def escape_query(text):
+def escape_query(text, chars_to_escape):
     """
     normalizes the query text to a format that can be consumed
     by the backend database
     """
     text = force_text(text)
     text = RE_SPACE.sub(" ", text)  # Standardize spacing.
-    text = RE_NON_WORD.sub(" ", text)  # Replace harmful characters with space.
+    text = chars_to_escape.sub(" ", text)  # Replace harmful characters with space.
```
Comment (etianen) [inline src/watson/backends.py]: It's not really `chars_to_escape`. It's more like `re_escape_chars`.

Nitpicky, but I'd expect `chars_to_escape` to be a string.

File: src/watson/backends.py [hunk 2]
```diff
@@ -174,10 +169,13 @@ class PostgresSearchBackend(SearchBackend):
 
     def escape_postgres_query(self, text):
         """Escapes the given text to become a valid ts_query."""
+        # PostgreSQL to_tsquery operators: ! & : ( ) |
+        chars_to_escape = re.compile(r'[&:(|)!]', re.UNICODE)
```
Comment (etianen) [inline src/watson/backends.py]: This regex is repeated twice in both postgres backends. We should move it to a global constant `RE_POSTGRES_ESCAPE_CHARS`. That way we also benefit from only compiling the regex once.

File: src/watson/backends.py [hunk 3]
```diff
@@ -322,10 +320,14 @@ class PostgresLegacySearchBackend(PostgresSearchBackend):
 
     def escape_postgres_query(self, text):
         """Escapes the given text to become a valid ts_query."""
+
+        # PostgreSQL to_tsquery operators: ! & : ( ) |
+        chars_to_escape = re.compile(r'[&:(|)!]', re.UNICODE)
```
Comment (etianen) [inline src/watson/backends.py]: As above, this should be a global constant `RE_POSTGRES_ESCAPE_CHARS`.

File: src/watson/backends.py [hunk 4]
```diff
@@ -341,11 +343,14 @@ class PostgresPrefixLegacySearchBackend(RegexSearchMixin, PostgresLegacySearchBa
 
 
 def escape_mysql_boolean_query(search_text):
+    # MySQL boolean full-text search operators: > < ( ) " ~ * + -
+    chars_to_escape = re.compile(r'["()><~*+-]', re.UNICODE)
```
Comment (etianen) [inline src/watson/backends.py]: A global constant `RE_MYSQL_ESCAPE_CHARS`.

**Comment by amureki** [comment]
@etianen thanks for review! All fine with your requests, I want to contribute good code, so it will be better for everyone. :)

I'll fix this as soon as possible, and update this PR.

**Comment by etianen** [comment]
Cool! Hit me with a message on this pull request when you're done,
otherwise I might not be notified of the changes. :)

**Comment by amureki** [comment]
@etianen done! :)

**Comment by etianen** [comment]
Thanks for sorting this!

I'll get a release out end of this week or start of next.

**Pull request merged**